tempesta_fw/ss_skb.c

```diff
@@ -1102,27 +1105,29 @@ struct sk_buff *
 ss_skb_split(struct sk_buff *skb, int len)
 {
 	struct sk_buff *buff;
-	int nsize, asize, nlen;
+	int n = 0;
 
 	/* Assert that the SKB is orphaned. */
 	WARN_ON_ONCE(skb->destructor);
 
-	nsize = skb_headlen(skb) - len;
-	if (nsize < 0)
-		nsize = 0;
-	asize = ALIGN(nsize, 4);
+	if (len < skb_headlen(skb))
+		n = skb_headlen(skb) - len;
 
-	buff = alloc_skb_fclone(asize + MAX_TCP_HEADER, GFP_ATOMIC);
-	if (buff == NULL)
+	buff = alloc_skb_fclone(ALIGN(n, 4) + MAX_TCP_HEADER, GFP_ATOMIC);
+	if (!buff)
 		return NULL;
 
 	skb_reserve(buff, MAX_TCP_HEADER);
-	/* Make sure there's exactly asize bytes available. */
-	buff->reserved_tailroom = buff->end - buff->tail - asize;
 
-	nlen = skb->len - len - nsize;
-	buff->truesize += nlen;
-	skb->truesize -= nlen;
+	n = skb->len - len;
+	buff->truesize += n;
```

Comment:
It seems that without `nsize` (`asize`) we have double `skb_headlen(skb) - len` part in `buff->truesize`, if `len < skb_headlen(skb)`: since `skb->truesize = SKB_TRUESIZE(size)` in `alloc_skb_fclone()`->`__alloc_skb_init()` and `size == asize` in this case; so, we already included `asize` into `skb->truesize` during `alloc_skb_fclone()` and we need to subtract it from `skb->len - len` value.